logrotate.c

```diff
@@ -1388,6 +1388,98 @@ static long daysElapsed(const struct tm *now, const struct tm *last)
     return (long) ((intmax_t)diff / DAY_SECONDS);
 }
 
+static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth);
+
+static int isSecureDirStep(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth)
+{
+    struct stat sb;
+
+    if (lstat(pathname, &sb) < 0)
+        return -errno;
+
+    if (S_ISLNK(sb.st_mode)) {
+        /* check whether symlink target is inside a secure directory */
+        char *canonpath;
+        int r;
+
+        canonpath = realpath(pathname, NULL);
+        if (!canonpath)
+            return -errno;
+
+        r = isSecureDir(canonpath, exempt_uid, exempt_gid, depth + 1);
+        free(canonpath);
+        return r;
+    }
+
+    if (!S_ISDIR(sb.st_mode))
+        return -ENOTDIR;
+
+    if ((sb.st_uid != ROOT_UID && sb.st_uid != exempt_uid) ||
+        ((sb.st_mode & S_IWGRP) && sb.st_gid != 0 && sb.st_gid != exempt_gid) ||
+        (sb.st_mode & S_IWOTH && !(sb.st_mode & S_ISVTX))) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth)
+{
+    char *copy, *p;
+    int r;
+
+    if (depth >= 10)
+        return -ELOOP;
+
+    if (pathname[0] != '/')
+        return -EINVAL;
+
+    copy = strdup(pathname);
+    if (!copy)
+        return -errno;
+
+    for (p = copy + 1; p; p = strchr(p + 1, '/')) {
+        char backup;
+
+        backup = *p;
+        *p = '\0';
+        r = isSecureDirStep(copy, exempt_uid, exempt_gid, depth);
```

Comment:
Could we simplify the loop such that `isSecureDirStep()` is called from a single place in the code?
```diff
--- a/logrotate.c
+++ b/logrotate.c
@@ -1438,20 +1438,20 @@ static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid,
     if (!copy)
         return -errno;

-    for (p = copy + 1; p; p = strchr(p + 1, '/')) {
+    for (p = copy + 1;; p = strchr(p + 1, '/')) {
         char backup;
+        if (p != NULL) {
+            backup = *p;
+            *p = '\0';
+        }

-        backup = *p;
-        *p = '\0';
         r = isSecureDirStep(copy, exempt_uid, exempt_gid, depth);
+        if (r <= 0 || p == NULL)
+            break;
+
         *p = backup;
-        if (r <= 0)
-            goto cleanup;
     }

-    r = isSecureDirStep(pathname, exempt_uid, exempt_gid, depth);
-
-cleanup:
     free(copy);
     return r;
 }
```

Reply:
Yes, I think so; adopted.